cores/esp32/esp32-hal-i2c.c

```diff
@@ -269,86 +283,90 @@ i2c_err_t i2cRead(i2c_t * i2c, uint16_t address, bool addr_10bit, uint8_t * data
     i2cResetCmd(i2c);
 
     //CMD START
-    i2cSetCmd(i2c, 0, I2C_CMD_RSTART, 0, false, false, false);
+    i2cSetCmd(i2c, cmdIdx++, I2C_CMD_RSTART, 0, false, false, false);
 
     //CMD WRITE ADDRESS
     i2c->dev->fifo_data.val = address & 0xFF;
     if(addr_10bit) {
         i2c->dev->fifo_data.val = (address >> 8) & 0xFF;
     }
-    i2cSetCmd(i2c, 1, I2C_CMD_WRITE, addrLen, false, false, true);
+    i2cSetCmd(i2c, cmdIdx++, I2C_CMD_WRITE, addrLen, false, false, true);
+    nextCmdCount = cmdIdx;
 
-    while(len) {
-        cmdIdx = (index)?0:2;
-        willRead = (len > 32)?32:(len-1);
-        if(cmdIdx){
-            i2cResetFiFo(i2c);
-        }
-
-        if(willRead){
-            i2cSetCmd(i2c, cmdIdx++, I2C_CMD_READ, willRead, false, false, false);
-        }
-
-        if((len - willRead) > 1) {
-            i2cSetCmd(i2c, cmdIdx++, I2C_CMD_END, 0, false, false, false);
-        } else {
-            willRead++;
-            i2cSetCmd(i2c, cmdIdx++, I2C_CMD_READ, 1, true, false, false);
-            if(sendStop) {
-                i2cSetCmd(i2c, cmdIdx++, I2C_CMD_STOP, 0, false, false, false);
-            }
-        }
-
-        //Clear Interrupts
-        i2c->dev->int_clr.val = 0xFFFFFFFF;
-
-        //START Transmission
-        i2c->dev->ctr.trans_start = 1;
+    //Clear Interrupts
+    i2c->dev->int_clr.val = 0x00001FFF;
 
+    //START Transmission
+    i2c->dev->ctr.trans_start = 1;
+    while (!stopped) {
         //WAIT Transmission
         uint32_t startAt = millis();
         while(1) {
             //have been looping for too long
-            if((millis() - startAt)>50){
-                log_e("Timeout! Addr: %x", address >> 1);
+            if((millis() - startAt)>50) {
```

Comment:
If you exit right here, the I2C State Machine is Where?  I think some Reset/Cleanup action need to be executed. The I2C protocol allows the Slave device to pause the Master by holding SCL low. I have not been able to determine the 'Official' maximum pause. Philips UM10204: pg.48 says 'Tlow min 4.7us, max = unlimited.  So, theoretically, a valid transaction could take until the death of the Universe!

Reply:
If we are exiting at that point, it is because the state machine has failed or the slave is not responding. The program needs to decide how to handle such as by calling i2cReset to reset the hardware state machine. The problem with the current ESP32 hardware implementation of  I2C is that it really doesn't support the slave holding the SCL low for very long at all. There is a timer in i2cInit that I have set to its longest value on line 455. I tried setting it longer with no additional benefit so it appears to be at its actual maximum. Numerous people have commented on this implementation flaw but the current version of the chip is what it is. In the IDF forums, making I2C compliant the spec is one of the big requests for the next chip revision.

Reply:
@lonerzzz Have you thought about address NAK, and Arbitration Lost errors?  If you setup a list of cmd ops:
 1. start
 2. write address (to select the address of the Slave)
 3. Read block of bytes
 4. Stop

And then you get a address NAK, does that cause the I2C StateMachine(SM)to halt? or does is continue with the READ command?  The Docs are very Fuzzy!
I was just thinking that if an Address NAK event happened, you should cleanup the CMD Stack and ReInit the SM back to an Idle condition. Then return the NAK error to the APP.

Reply:
Oh, yeah, the docs need some work. When I finally got through to the espressif folks a week ago, they indicated they were working on I2C code fixes as well as documentation and would like my fix to include so I was trying to get it done amongst other tasks. The good news is at least we know some more improvements are coming. They are definitely needed.

To your question. my understanding is that most errors do cause the SM to halt as one would expect but I have seen situations where that was not the case but 'reproducability' was the biggest obstacle to getting to the root of it.

You are correct. We could clean up on the NAK if we can confirm that the SM is stable enough to properly stop the pipeline. My assumption so far has been that in these situations, I let the application layer call i2cReset and reset the whole state machine simply because I am not sure of what all the problems with the SM are.